[css-anchor-position-1] Minimize computed serialization of position-area

tuankiet65 opened the pull request:
#### 436e14dbc03673a12b20d0679567ab0ea8b8fdae
<pre>
[css-anchor-position-1] Minimize computed serialization of position-area
<a href="https://rdar.apple.com/148529924">rdar://148529924</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=291004">https://bugs.webkit.org/show_bug.cgi?id=291004</a>

Reviewed by Tim Nguyen.

Change the computed value serialization of position-area to match the
spec: <a href="https://drafts.csswg.org/css-anchor-position-1/#position-area-computed">https://drafts.csswg.org/css-anchor-position-1/#position-area-computed</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-computed.html:
    - Rewrite the test to test all possible valid values of position-area

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-parsing.html:
    - Add some valid values that weren&apos;t tested (self-block/inline-start/end keywords)

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-position-area-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-position-area.html:
    - Update expected computed value of position-area to follow the resolution.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp:
(WebCore::CSSPropertyParserHelpers::makeAmbiguous):
    - Add function to make axis-unambiguous keywords ambiguous
      (e.g block-start =&gt; start). This is used to minimize the
      computed value of position-area in contexts that allows it.

(WebCore::CSSPropertyParserHelpers::valueForPositionArea):
(WebCore::CSSPropertyParserHelpers::consumePositionArea):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.h:
* Source/WebCore/style/StyleExtractorConverter.h:
(WebCore::Style::ExtractorConverter::convertPositionArea):

Canonical link: <a href="https://commits.webkit.org/297894@main">https://commits.webkit.org/297894@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6034d321d3ef6dde138cc4ba5352a9e1dcf02d32

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113262 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33002 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23475 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119470 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63928 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115224 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33650 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41560 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86200 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116209 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26879 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101882 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66525 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26143 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20010 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63224 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96253 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20085 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122687 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40340 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95048 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40725 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98097 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94793 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24195 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39939 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17757 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36476 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40226 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45725 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39867 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43200 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41604 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fantasai commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.h:
```diff
@@ -41,14 +41,29 @@ struct PropertyParserState;
 
+enum class PositionAreaContext {
+    Serialization,
+    Computed
```
Should this be `Specified` instead of `Serialization`?

tuankiet65 replied:
Yeah it should be `Specified`, the words are similar enough that I mix them up 😛

fantasai reviewed (commented):
These tests are really thorough. @_@

r=fantasai

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp:
```diff
@@ -316,7 +362,7 @@ RefPtr<CSSValue> consumePositionArea(CSSParserTokenRange& range, CSS::PropertyPa
     auto dim2 = *maybeDim2;
 
-    return valueForPositionArea(dim1, dim2);
+    return minimizePositionArea(dim1, dim2, PositionAreaContext::Specified);
```
Specified and computed should use the same shortest serialization, we usually apply shortest serialization to both.

https://drafts.csswg.org/cssom/#serialize-a-css-value

tuankiet65 replied:
I had two different paths because it would require changes to the parsing test, which I don't want to do. But from your comment, I'd do it (minimize serialization for both specified and computed value)

fantasai replied:
The spec is actually not entirely clear on this point--it only defines the conversion for computed values--so I suggest we land this and then deal with the specified value serialization question later.

tuankiet65 replied:
Resolved to land this.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp:
```diff
@@ -316,7 +364,7 @@ RefPtr<CSSValue> consumePositionArea(CSSParserTokenRange& range, CSS::PropertyPa
     auto dim2 = *maybeDim2;
 
-    return valueForPositionArea(dim1, dim2);
+    return minimizePositionArea(dim1, dim2, PositionAreaContext::Specified);
```
I prefer the old name `valueForPositionArea`. `minimize` adds confusion through inconsistency with the rest of the code IMO

tuankiet65 replied:
I changed it back to `valueForPositionArea`

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.h:
```diff
@@ -41,14 +41,29 @@ struct PropertyParserState;
 
 namespace CSSPropertyParserHelpers {
 
+enum class PositionAreaContext {
```
```suggestion
enum class ValueType {
```

tuankiet65 replied:
I dunno about this, `WebCore::CSSPropertyParserHelpers::ValueType` sounds a bit too generic?

nt1m replied:
That's what it represents though.

tuankiet65 replied:
🤔 I guess if we make it generic then it should live in some more generic sounding header file but okay

tuankiet65 replied:
Changed to `ValueType`

nt1m reviewed (approved):
r=me with nits addressed

tuankiet65 committed:
[css-anchor-position-1] Minimize computed serialization of position-area
rdar://148529924
https://bugs.webkit.org/show_bug.cgi?id=291004

Reviewed by Tim Nguyen.

Change the computed value serialization of position-area to match the
spec: https://drafts.csswg.org/css-anchor-position-1/#position-area-computed

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-computed.html:
    - Rewrite the test to test all possible valid values of position-area

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-parsing.html:
    - Add some valid values that weren't tested (self-block/inline-start/end keywords)

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-position-area-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-position-area.html:
    - Update expected computed value of position-area to follow the resolution.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp:
(WebCore::CSSPropertyParserHelpers::makeAmbiguous):
    - Add function to make axis-unambiguous keywords ambiguous
      (e.g block-start => start). This is used to minimize the
      computed value of position-area in contexts that allows it.

(WebCore::CSSPropertyParserHelpers::valueForPositionArea):
(WebCore::CSSPropertyParserHelpers::consumePositionArea):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.h:
* Source/WebCore/style/StyleExtractorConverter.h:
(WebCore::Style::ExtractorConverter::convertPositionArea):

Canonical link: https://commits.webkit.org/297894@main
